### ISSUE TITLE
Way to support or ignore CSS-in-JS logic

### ISSUE THREAD

**Issue by wfortin** [AndyOGo/stylelint-declaration-strict-value#134]
I would like to be able to suppress errors when using CSS-in-JS logic.

Example:
```jsx
export const MyButton = styled.button`
    font-size: ${({inherit}) => (inherit ? 'inherit' : tokens.fontMedium)};
`
```

With this ruleset:
```
"scale-unlimited/declaration-strict-value": "font-size"
```

I get the error:
```
Expected variable, function or keyword for "=>" of "font-size"       scale-unlimited/declaration-strict-value
```

Extracting to a named function works:
```jsx
const getFontSize = ({inherit}) => (inherit ? 'inherit' : tokens.fontMedium)

export const MyButton = styled.button`
    font-size: ${getFontSize};
`
```

Would it be possible to ignore or allow any inline JS logic sections in CSS-in-JS styles? If this could be included with "ignoreFunctions: true" that would be great!

**Comment by wfortin** [comment]
Also adding `"=>"` as an ignored value doesn't fix the error, it now says:

```
Expected variable, function or keyword for "(inherit ? 'inherit' : tokens.fontMedium)}" of "font-size"           scale-unlimited/declaration-strict-value
```

**Comment by AndyOGo** [comment]
Hi @wfortin 

Thanks for your issue.

CSS-in-JS isn't supported.

Please have a look at the readme for supported languages:
> A [stylelint](https://github.com/stylelint/stylelint) plugin that enforces either variables (`$sass`, `namespace.$sass`, `@less`, `var(--cssnext)`, `css-loader @value`), functions or custom CSS values, like keywords (`inherit`, `none`, `currentColor` etc.), colors (`#fff`, `red`, etc.) or numbers incl. units (`0`, `1px`, `100%`, etc.) for CSS longhand and shorthand properties.

**Comment by wfortin** [comment]
I know it's not officially supported but still works in most of the cases we need it fortunately.

**Comment by AndyOGo** [comment]
Okay. I assume it works until you use JavaScript (the fat arrow above).

This plugin works with CSS AST nodes, it's not aware of ES AST nodes.

**Comment by wfortin** [comment]
We have a mix of CSS modules and are slowly migrating to styled-components. Here's our stylelint config:

```js
module.exports = {
    extends: [
        'stylelint-config-standard',
        'stylelint-config-prettier',
        'stylelint-config-css-modules'
    ],
    plugins: [
        'stylelint-order',
        'stylelint-config-rational-order/plugin',
        'stylelint-declaration-strict-value'
    ],
    rules: {
        indentation: null,
        'rule-empty-line-before': [
            'always',
            {
                except: ['after-single-line-comment', 'first-nested']
            }
        ],
        'at-rule-no-unknown': [
            true,
            {
                ignoreAtRules: ['define-mixin', 'for', 'mixin']
            }
        ],
        'selector-pseudo-class-no-unknown': [
            true,
            {
                ignorePseudoClasses: ['global']
            }
        ],
        'function-calc-no-unspaced-operator': null,
        'declaration-block-single-line-max-declarations': null,
        'font-family-no-missing-generic-family-keyword': null,
        'declaration-empty-line-before': null,
        'function-name-case': null,
        'value-keyword-case': null,
        'function-whitespace-after': null,
        'color-hex-length': 'long',
        'order/properties-order': [
            [],
            {
                severity: 'warning'
            }
        ],
        'plugin/rational-order': [
            true,
            {
                'border-in-box-model': false,
                'empty-line-between-groups': false,
                severity: 'warning'
            }
        ],
        'scale-unlimited/declaration-strict-value': [
            [
                '/margin-/',
                '/padding-/',
                '/color/',
                'background',
                'background-color',
                'opacity',
                'border-radius',
                'border-top-left-radius',
                'border-top-right-radius',
                'border-bottom-right-radius',
                'border-bottom-left-radius',
                'font-weight',
                'box-shadow',
                'font-size',
                'z-index',
                'transition',
                'transition-duration',
                'transition-delay',
                'line-height'
            ],
            {
                ignoreValues: {
                    '/margin-/': ['auto', '0'],
                    '/padding-/': ['auto', '0'],
                    '/color/': ['currentColor', 'inherit', 'transparent', '/theme.*/'],
                    background: ['inherit', 'transparent', 'none', '/theme.*/'],
                    'background-color': ['inherit', 'transparent', '/theme.*/'],
                    opacity: ['0', '1'],
                    'border-radius': ['inherit', '0'],
                    'box-shadow': ['none'],
                    'font-weight': ['inherit'],
                    'font-size': [
                        '0',
                        '10px',
                        '12px',
                        '14px',
                        '16px',
                        '18px',
                        '20px',
                        '24px',
                        '32px',
                        '48px',
                        'inherit'
                    ],
                    'line-height': ['0', '1', 'inherit'],
                    'border-top-left-radius': ['0'],
                    'border-top-right-radius': ['0'],
                    'border-bottom-right-radius': ['0'],
                    'border-bottom-left-radius': ['0']
                },
                disableFix: true
            }
        ]
    }
}
```

**Comment by AndyOGo** [comment]
Thank you.

Well CSS-in-JS basically allows for any ES expression.
How would you solve that, considering that this plugin only supports 3 CSS-like syntaxes for CSS variables, function and values?

**Comment by wfortin** [comment]
The quick fix for this would be to consider that anything inside the `${}` would be functions. The with `ignoreFunctions` (that I believe to be true by default) we wouldn't get any error. I don't know how feasible is this.

**Comment by AndyOGo** [comment]
Well, here is the AST for just that single case you posted.

https://astexplorer.net/#/gist/8f43920cf2040735c0f1d59c89923eda/53806cd063e0784a880a8345eaf8d2378a2e18ce

How would a proper solution look, supporting any expressions ECMAScript provides?

**Comment by wfortin** [comment]
That goes beyond my knowlegdge of stylelint, I actually wasn't aware that stylelint supports parsing the AST...

**Comment by AndyOGo** [comment]
No worries. You can learn it, there are plenty of free online resources for handling ASTs.
You are welcome to provide a PR.
